src/rs_distributions/transforms/fill_scale_tril.py

```diff
@@ -0,0 +1,84 @@
+import torch
+from torch.distributions import Transform, constraints
+from torch.distributions.transforms import SoftplusTransform
+from torch.distributions.utils import vec_to_tril_matrix, tril_matrix_to_vec
+
+
+class FillScaleTriL(Transform):
+    def __init__(self, diag_transform=None, diag_shift=1e-05):
+        """
+        Converts a tensor into a lower triangular matrix with positive diagonal entries.
+
+        Args:
+            diag_transform: transformation used on diagonal to ensure positive values.
+            Default is SoftplusTransform
+            diag_shift (float): small offset to avoid diagonals very close to zero.
+            Default offset is 1e-05
+
+        """
+        super().__init__()
+        self.diag_transform = (
+            diag_transform if diag_transform is not None else SoftplusTransform()
+        )
+        self.diag_shift = diag_shift
+
+        domain = constraints.real_vector
+        codomain = constraints.lower_cholesky
+        bijective = True
+
+    def _call(self, x):
+        """
+        Transform input vector to lower triangular.
+
+        Args:
+            x (torch.Tensor): Input vector to transform
+        Returns:
+            torch.Tensor: Transformed lower triangular matrix
+        """
+        x = vec_to_tril_matrix(x)
+        diagonal_elements = x.diagonal(dim1=-2, dim2=-1)
+        transformed_diagonal = self.diag_transform(diagonal_elements)
+        if self.diag_shift is not None:
+            transformed_diagonal += self.diag_shift
+        x.diagonal(dim1=-2, dim2=-1).copy_(transformed_diagonal)
+        return x
+
+    def _inverse(self, y):
+        """
+        Apply the inverse transformation to the input lower triangular matrix.
+
+        Args:
+            y (torch.Tensor): Invertible lower triangular matrix
+
+        Returns:
+            torch.Tensor: Inversely transformed vector
+
+        """
+        diagonal_elements = y.diagonal(dim1=-2, dim2=-1)
+        if self.diag_shift is not None:
+            transformed_diagonal = self.diag_transform.inv(
+                diagonal_elements - self.diag_shift
+            )
+        else:
+            transformed_diagonal = self.diag_transform.inv(diagonal_elements)
+        y.diagonal(dim1=-2, dim2=-1).copy_(transformed_diagonal)
+        return tril_matrix_to_vec(y)
+
+    def log_abs_det_jacobian(self, x, y):
+        """
+        Computes the log absolute determinant of the Jacobian matrix for the transformation.
+
+        Assumes that Softplus is used on the diagonal.
```

Comment:
Okay I will modify to get derivatives of `self.diag_transform`

Reply:
Im not entirely sure if this is the correct way to do it, but this is what I have so far: 

```python 

    def log_abs_det_jacobian(self, x, y):
        L = vec_to_tril_matrix(x)
        diag = L.diagonal()
        diag.requires_grad_(True)
        transformed_diag = self.diag_transform(diag + self.diag_shift)
        derivatives = torch.autograd.grad(
            outputs=transformed_diag,
            inputs=diag,
            grad_outputs=torch.ones_like(transformed_diag),
        )[0]
        log_det_jacobian = torch.log(torch.abs(derivatives)).sum()
        return log_det_jacobian
```
When I use `SoftplusTransform` for `self.diag_transform` this method matches the gradients if I use: 

```python 
transform = FillScaleTriL()
softplus_transform = SoftplusTransform()

x = torch.randn(3, requires_grad=True)
L = transform._call(x)

# using analytical form
L_ = vec_to_tril_matrix(x)
diag = L_.diagonal()
derivatives = torch.sigmoid(diag)
result1 = torch.log(torch.abs(derivatives)).sum()

# using FillScaleTriL.log_abs_det_jacobian
result2 = transform.log_abs_det_jacobian(x,L)

#checking if gradients match
torch.allclose(result1,result2)
```

Reply:
I have some thoughts about this. In tfp, the `FillScaleTriL` bijector is [implemented](https://github.com/tensorflow/probability/blob/65f265c62bb1e2d15ef3e25104afb245a6d52429/tensorflow_probability/python/bijectors/fill_scale_tril.py#L132) as a `Chain` which is their equivalent of the `ComposeTransform` class in `torch`.  Specifically, it is the composition of two bijectors, `FillTriL` and `TransformDiagonal`. Here are the Jacobian implementations for those base classes:
 - [FillTriL](https://github.com/tensorflow/probability/blob/65f265c62bb1e2d15ef3e25104afb245a6d52429/tensorflow_probability/python/bijectors/fill_triangular.py#L97)
 - [TransformDiagonal](https://github.com/tensorflow/probability/blob/65f265c62bb1e2d15ef3e25104afb245a6d52429/tensorflow_probability/python/bijectors/transform_diagonal.py#L81)
 The `FillTriL` is pretty trivial because it essentially amounts to reshaping the input in a funny way. The TransformDiagonal Jacobian is __very__ straightforward to implement but less obvious. The developer left a nice note explaining it. 
 
 My suggestion is that you adopt this structure for your own implementation. Write your own `FillTriL` and `TransformDiagonal` transforms and implement `FillScaleTriL` as a `ComposeTransform`